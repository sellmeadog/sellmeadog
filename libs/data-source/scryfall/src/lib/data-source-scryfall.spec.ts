import { makeMockedFunction } from '@sellmeadog/extensions/testing';
import { createReadStream } from 'fs';
import fetch, { Response } from 'node-fetch';
import { resolve } from 'path';
import { lastValueFrom } from 'rxjs';
import {
  make_scryfall_data_source,
  ScryfallDataSource,
} from './data-source-scryfall';

describe('ScryfallDataSource', () => {
  let fetch_: jest.MockedFunction<typeof fetch>;
  let scryfall: ScryfallDataSource;

  beforeEach(() => {
    jest.resetAllMocks();

    fetch_ = makeMockedFunction(fetch);
    scryfall = new ScryfallDataSource();
  });

  it('factory works', () => {
    expect(make_scryfall_data_source()).toBeInstanceOf(ScryfallDataSource);
  });

  it('should GET /catalog endpoints', async () => {
    fetch_.mockResolvedValue(
      new Response(
        createReadStream(
          resolve(__dirname, '__json__', 'catalog-keyword-actions.json')
        )
      )
    );

    const data = await lastValueFrom(scryfall.catalog('keyword-actions'));

    expect(data).toMatchSnapshot();
    expect(fetch_).toHaveBeenCalledTimes(1);
    expect(fetch_).toHaveBeenCalledWith(
      'https://api.scryfall.com/catalog/keyword-actions'
    );
  });

  it('should GET /bulk-data/default_cards endpoint', async () => {
    const download_uri = resolve(
      __dirname,
      '__json__',
      'bulk-data-oracle-cards.json'
    );

    fetch_
      .mockImplementationOnce(() =>
        Promise.resolve(new Response(JSON.stringify({ download_uri })))
      )
      .mockImplementationOnce((url) =>
        Promise.resolve(new Response(createReadStream(url.toString())))
      );

    const data = await lastValueFrom(scryfall.default_cards());

    expect(data).toMatchSnapshot();
    expect(fetch_).toHaveBeenCalledTimes(2);
    expect(fetch_).toHaveBeenCalledWith(
      'https://api.scryfall.com/bulk-data/default_cards'
    );
    expect(fetch_).toHaveBeenCalledWith(download_uri);
  });

  it('should GET /bulk-data/oracle_cards endpoint', async () => {
    const download_uri = resolve(
      __dirname,
      '__json__',
      'bulk-data-oracle-cards.json'
    );

    fetch_
      .mockImplementationOnce(() =>
        Promise.resolve(new Response(JSON.stringify({ download_uri })))
      )
      .mockImplementationOnce((url) =>
        Promise.resolve(new Response(createReadStream(url.toString())))
      );

    const data = await lastValueFrom(scryfall.oracle_cards());

    expect(data).toMatchSnapshot();
    expect(fetch_).toHaveBeenCalledTimes(2);
    expect(fetch_).toHaveBeenCalledWith(
      'https://api.scryfall.com/bulk-data/oracle_cards'
    );
    expect(fetch_).toHaveBeenCalledWith(download_uri);
  });
});
