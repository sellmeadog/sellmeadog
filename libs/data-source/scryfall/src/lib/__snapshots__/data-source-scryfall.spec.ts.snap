// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScryfallDataSource should GET /bulk-data/default_cards endpoint 1`] = `
Object {
  "artist": "John Matson",
  "artist_ids": Array [
    "a1685587-4b55-446b-b420-c37b202ed3f1",
  ],
  "booster": true,
  "border_color": "white",
  "card_back_id": "0aeebaf5-8c7d-4636-9e82-8c27447861f7",
  "cardmarket_id": 12551,
  "cmc": 2,
  "collector_number": "100",
  "color_identity": Array [
    "U",
  ],
  "colors": Array [
    "U",
  ],
  "digital": false,
  "edhrec_rank": 12284,
  "finishes": Array [
    "nonfoil",
    "foil",
  ],
  "flavor_text": "Storm crow descending, winter unending. Storm crow departing, summer is starting.",
  "foil": true,
  "frame": "2003",
  "full_art": false,
  "games": Array [
    "paper",
    "mtgo",
  ],
  "highres_image": true,
  "id": "036ef8c9-72ac-46ce-af07-83b79d736538",
  "illustration_id": "d01aa92b-0582-4e1e-a7b0-737b2ad4e462",
  "image_status": "highres_scan",
  "image_uris": Object {
    "art_crop": "https://c1.scryfall.com/file/scryfall-cards/art_crop/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "border_crop": "https://c1.scryfall.com/file/scryfall-cards/border_crop/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "large": "https://c1.scryfall.com/file/scryfall-cards/large/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "normal": "https://c1.scryfall.com/file/scryfall-cards/normal/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "png": "https://c1.scryfall.com/file/scryfall-cards/png/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.png?1562730661",
    "small": "https://c1.scryfall.com/file/scryfall-cards/small/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
  },
  "keywords": Array [
    "Flying",
  ],
  "lang": "en",
  "layout": "normal",
  "legalities": Object {
    "alchemy": "not_legal",
    "brawl": "not_legal",
    "commander": "legal",
    "duel": "legal",
    "future": "not_legal",
    "gladiator": "not_legal",
    "historic": "not_legal",
    "historicbrawl": "not_legal",
    "legacy": "legal",
    "modern": "legal",
    "oldschool": "not_legal",
    "pauper": "legal",
    "paupercommander": "legal",
    "penny": "legal",
    "pioneer": "not_legal",
    "premodern": "legal",
    "standard": "not_legal",
    "vintage": "legal",
  },
  "mana_cost": "{1}{U}",
  "mtgo_foil_id": 22610,
  "mtgo_id": 22609,
  "multiverse_ids": Array [
    83282,
  ],
  "name": "Storm Crow",
  "nonfoil": true,
  "object": "card",
  "oracle_id": "000d5588-5a4c-434e-988d-396632ade42c",
  "oracle_text": "Flying (This creature can't be blocked except by creatures with flying or reach.)",
  "oversized": false,
  "power": "1",
  "prices": Object {
    "eur": "0.02",
    "eur_foil": "1.89",
    "tix": "0.03",
    "usd": "0.18",
    "usd_etched": null,
    "usd_foil": "2.49",
  },
  "prints_search_uri": "https://api.scryfall.com/cards/search?order=released&q=oracleid%3A000d5588-5a4c-434e-988d-396632ade42c&unique=prints",
  "promo": false,
  "rarity": "common",
  "related_uris": Object {
    "edhrec": "https://edhrec.com/route/?cc=Storm+Crow",
    "gatherer": "https://gatherer.wizards.com/Pages/Card/Details.aspx?multiverseid=83282",
    "mtgtop8": "https://mtgtop8.com/search?MD_check=1&SB_check=1&cards=Storm+Crow",
    "tcgplayer_infinite_articles": "https://infinite.tcgplayer.com/search?contentMode=article&game=magic&partner=scryfall&q=Storm+Crow&utm_campaign=affiliate&utm_medium=api&utm_source=scryfall",
    "tcgplayer_infinite_decks": "https://infinite.tcgplayer.com/search?contentMode=deck&game=magic&partner=scryfall&q=Storm+Crow&utm_campaign=affiliate&utm_medium=api&utm_source=scryfall",
  },
  "released_at": "2005-07-29",
  "reprint": true,
  "reserved": false,
  "rulings_uri": "https://api.scryfall.com/cards/036ef8c9-72ac-46ce-af07-83b79d736538/rulings",
  "scryfall_set_uri": "https://scryfall.com/sets/9ed?utm_source=api",
  "scryfall_uri": "https://scryfall.com/card/9ed/100/storm-crow?utm_source=api",
  "set": "9ed",
  "set_id": "e70c8572-4732-4e92-a140-b4e3c1c84c93",
  "set_name": "Ninth Edition",
  "set_search_uri": "https://api.scryfall.com/cards/search?order=set&q=e%3A9ed&unique=prints",
  "set_type": "core",
  "set_uri": "https://api.scryfall.com/sets/e70c8572-4732-4e92-a140-b4e3c1c84c93",
  "story_spotlight": false,
  "tcgplayer_id": 12835,
  "textless": false,
  "toughness": "2",
  "type_line": "Creature — Bird",
  "uri": "https://api.scryfall.com/cards/036ef8c9-72ac-46ce-af07-83b79d736538",
  "variation": false,
}
`;

exports[`ScryfallDataSource should GET /bulk-data/oracle_cards endpoint 1`] = `
Object {
  "artist": "John Matson",
  "artist_ids": Array [
    "a1685587-4b55-446b-b420-c37b202ed3f1",
  ],
  "booster": true,
  "border_color": "white",
  "card_back_id": "0aeebaf5-8c7d-4636-9e82-8c27447861f7",
  "cardmarket_id": 12551,
  "cmc": 2,
  "collector_number": "100",
  "color_identity": Array [
    "U",
  ],
  "colors": Array [
    "U",
  ],
  "digital": false,
  "edhrec_rank": 12284,
  "finishes": Array [
    "nonfoil",
    "foil",
  ],
  "flavor_text": "Storm crow descending, winter unending. Storm crow departing, summer is starting.",
  "foil": true,
  "frame": "2003",
  "full_art": false,
  "games": Array [
    "paper",
    "mtgo",
  ],
  "highres_image": true,
  "id": "036ef8c9-72ac-46ce-af07-83b79d736538",
  "illustration_id": "d01aa92b-0582-4e1e-a7b0-737b2ad4e462",
  "image_status": "highres_scan",
  "image_uris": Object {
    "art_crop": "https://c1.scryfall.com/file/scryfall-cards/art_crop/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "border_crop": "https://c1.scryfall.com/file/scryfall-cards/border_crop/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "large": "https://c1.scryfall.com/file/scryfall-cards/large/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "normal": "https://c1.scryfall.com/file/scryfall-cards/normal/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
    "png": "https://c1.scryfall.com/file/scryfall-cards/png/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.png?1562730661",
    "small": "https://c1.scryfall.com/file/scryfall-cards/small/front/0/3/036ef8c9-72ac-46ce-af07-83b79d736538.jpg?1562730661",
  },
  "keywords": Array [
    "Flying",
  ],
  "lang": "en",
  "layout": "normal",
  "legalities": Object {
    "alchemy": "not_legal",
    "brawl": "not_legal",
    "commander": "legal",
    "duel": "legal",
    "future": "not_legal",
    "gladiator": "not_legal",
    "historic": "not_legal",
    "historicbrawl": "not_legal",
    "legacy": "legal",
    "modern": "legal",
    "oldschool": "not_legal",
    "pauper": "legal",
    "paupercommander": "legal",
    "penny": "legal",
    "pioneer": "not_legal",
    "premodern": "legal",
    "standard": "not_legal",
    "vintage": "legal",
  },
  "mana_cost": "{1}{U}",
  "mtgo_foil_id": 22610,
  "mtgo_id": 22609,
  "multiverse_ids": Array [
    83282,
  ],
  "name": "Storm Crow",
  "nonfoil": true,
  "object": "card",
  "oracle_id": "000d5588-5a4c-434e-988d-396632ade42c",
  "oracle_text": "Flying (This creature can't be blocked except by creatures with flying or reach.)",
  "oversized": false,
  "power": "1",
  "prices": Object {
    "eur": "0.02",
    "eur_foil": "1.89",
    "tix": "0.03",
    "usd": "0.18",
    "usd_etched": null,
    "usd_foil": "2.49",
  },
  "prints_search_uri": "https://api.scryfall.com/cards/search?order=released&q=oracleid%3A000d5588-5a4c-434e-988d-396632ade42c&unique=prints",
  "promo": false,
  "rarity": "common",
  "related_uris": Object {
    "edhrec": "https://edhrec.com/route/?cc=Storm+Crow",
    "gatherer": "https://gatherer.wizards.com/Pages/Card/Details.aspx?multiverseid=83282",
    "mtgtop8": "https://mtgtop8.com/search?MD_check=1&SB_check=1&cards=Storm+Crow",
    "tcgplayer_infinite_articles": "https://infinite.tcgplayer.com/search?contentMode=article&game=magic&partner=scryfall&q=Storm+Crow&utm_campaign=affiliate&utm_medium=api&utm_source=scryfall",
    "tcgplayer_infinite_decks": "https://infinite.tcgplayer.com/search?contentMode=deck&game=magic&partner=scryfall&q=Storm+Crow&utm_campaign=affiliate&utm_medium=api&utm_source=scryfall",
  },
  "released_at": "2005-07-29",
  "reprint": true,
  "reserved": false,
  "rulings_uri": "https://api.scryfall.com/cards/036ef8c9-72ac-46ce-af07-83b79d736538/rulings",
  "scryfall_set_uri": "https://scryfall.com/sets/9ed?utm_source=api",
  "scryfall_uri": "https://scryfall.com/card/9ed/100/storm-crow?utm_source=api",
  "set": "9ed",
  "set_id": "e70c8572-4732-4e92-a140-b4e3c1c84c93",
  "set_name": "Ninth Edition",
  "set_search_uri": "https://api.scryfall.com/cards/search?order=set&q=e%3A9ed&unique=prints",
  "set_type": "core",
  "set_uri": "https://api.scryfall.com/sets/e70c8572-4732-4e92-a140-b4e3c1c84c93",
  "story_spotlight": false,
  "tcgplayer_id": 12835,
  "textless": false,
  "toughness": "2",
  "type_line": "Creature — Bird",
  "uri": "https://api.scryfall.com/cards/036ef8c9-72ac-46ce-af07-83b79d736538",
  "variation": false,
}
`;

exports[`ScryfallDataSource should GET /catalog endpoints 1`] = `
Array [
  "Meld",
  "Bolster",
  "Clash",
  "Fateseal",
  "Manifest",
  "Monstrosity",
  "Populate",
  "Proliferate",
  "Scry",
  "Support",
  "Detain",
  "Explore",
  "Fight",
  "Amass",
  "Adapt",
  "Assemble",
  "Abandon",
  "Activate",
  "Attach",
  "Seek",
  "Cast",
  "Counter",
  "Create",
  "Destroy",
  "Discard",
  "Double",
  "Exchange",
  "Exile",
  "Investigate",
  "Play",
  "Regenerate",
  "Reveal",
  "Sacrifice",
  "Set in motion",
  "Shuffle",
  "Tap",
  "Untap",
  "Vote",
  "Transform",
  "Surveil",
  "Goad",
  "Planeswalk",
  "Mill",
  "Learn",
  "Conjure",
  "Exert",
]
`;
